server: retry ambiguous kv/sql operations on node startup

aliher1911 opened the pull request:
Previously, nodes could fail on startup if node itself was needed to restore quorum on ranges. This problem was caused by circuit breaker failing fast on live node when talking back to restarting node instead of waiting for raft operation to run again (and succeed).
This PR adds explicit retries and also adds assertions enabled in race build that would fire if startup reaches dist sender or internal sql executor without wrapping it in startup retry helper.

Release note: None

Fixes #74714

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97710)
<!-- Reviewable:end -->

aliher1911 commented:
Alternative approach which "magically" invokes retry https://github.com/cockroachdb/cockroach/pull/98129 is less invasive but has context affecting underlying call behaviours.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,147 @@
+
+// RunWithStartupRetry synchronously runs a function with retry of circuit
+// breaker errors on startup.
+func RunWithStartupRetry(
```
I added a commit that does this.

tbg replied:
name nit: `startup.Retry` and `startup.RetryErr` might be a bit more catchy but up to you.

aliher1911 replied:
Good idea with [struct{}], generic usage patterns didn't fully settle. It would be fun to backport that as generics will have to be removed, but I think it is worth doing it with generics on master and without on backport.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,147 @@
+var runningStartup atomic.Int32
+
+// RetryContext adds startup tag to allow kv access from startup path.
+func RetryContext(ctx context.Context) context.Context {
```
Name suggestion: `WithoutChecks`

to make clear that you get a context that avoids all of the checks in this package. I.e. you opt into knowing better than the assertions.
A clearer comment to that effect would also be helpful.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,147 @@
+		if ctx.Err() != nil || !IsRetryableReplicaError(err) {
+			break
+		}
+		log.VEventf(ctx, 1, "failed %s during node startup, retrying %s", opName, err)
```
`Infof`? At least if we've been trying for a few seconds.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,147 @@
+// e.g. code paths that avoid asynchronous execution or creating jobs and
+// instead run eagerly during startup. This shouldn't be used to suppress
+// startup assertions as makes server vulnerable to crash loops on startup.
+func RunWithoutRetry(ctx context.Context, f func(ctx context.Context) error) error {
```
I don't think we need this - if you have a `var f func(ctx context.Context) X` you can also just invoke it with `f(WithoutChecks(ctx))`. Also, there's only one caller.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,147 @@
+	return f(ctx)
+}
+
+func MaybeTagStartupContext(ctx context.Context) func() {
```
This doesn't do anything to the `ctx`, let's clean this up. How about

```go
done := startup.Begin()
defer done()
```

It can still take a `ctx` if we can do something useful with it.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,147 @@
+	}
+}
+
+func AssertStartupRetry(ctx context.Context) {
```
This seems like it could have false positives. You want to check that while starting up, all goroutines that call this have a tagged context. But when startup is "done", and there is a slight delay to actually "unregister" the startup, can't there be something in the system that doesn't have the correctly tagged ctx just because it has very little to do with the startup procedure?

tbg commented on pkg/server/server_int_test.go:
```diff
@@ -0,0 +1,354 @@
+		ServerArgsPerNode: make(map[int]base.TestServerArgs),
+		ReusableListeners: true,
+	}
+	var enableFaults atomic.Bool
```
is this WIP? This atomic is never queried. Is the idea in these tests to start a node against a cluster where some/all ranges have circuit breakers enabled? That is nice! Let me know if you want to figure out how to do this, I'm unsure of whether these tests are up for review or not.

tbg reviewed (commented):
First coarse review, I didn't go to deep since it's still a draft.

The approach makes sense - I'm not sure we can do the assertions like you have them here, which is a bit sad, hopefully I'm missing something.

Pushing a small fixup commit to inline something, just to save you the typing.

I've looked at the other approach too - I don't know, there just isn't going to be a super pretty solution, I somewhat prefer the approach you have here, at least it is explicit and we can document it well.

tbg commented on pkg/server/server_startup_test.go:
```diff
@@ -0,0 +1,310 @@
+// If this test failed, then suspicious range start key deduced from error could
+// be used in TestStartupFailure to investigate code path not covered by startup
+// retries.
+func TestStartupFailureRandomRange(t *testing.T) {
```
```
--- PASS: TestStartupFailureRandomRange (31.04s)
```

Spending 30+s on a single test seems difficult to justify, especially since we also have `TestStartupInjectedFailureSingleNode` which takes less than 2s and which should provide similar coverage.

I'm even hesitant to merge this for only nightly stress, since spending 30s on each iteration of the package on this test cannibalizes other tests and reduces our chances of finding flakes.

We could move this test into a leaf package (`startup_test`, i.e. in `startup`?) to avoid this problem - the `startup` package would basically be stressing these new tests only. Is there a dependency that would prohibit this (i.e. are you using anything in `server_test` in this test)? Note that this test should continue to run only under stress; we don't want a 30+s test on PR/bors CI.

tbg commented on pkg/server/server_startup_test.go:
```diff
@@ -0,0 +1,310 @@
+				SpanConfig: &spanconfig.TestingKnobs{
+					ConfigureScratchRange: true,
+				},
+				Store: &kvserver.StoreTestingKnobs{
```
Nice test.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,190 @@
+// startup.
+//
+// Server startup is now heavily dependent on kv for retrieving various
+// metadata. With introduction of circuit breakers, it is now possible for
```
This isn't new with circuit breakers: you could always get AmbiguousResultError and you needed to be ready to handle them. But of course circuit breakers can make this failure mode very, very sticky and so it now presents a much larger problem.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,190 @@
+// then the node would not be able to rejoin the cluster after circuit breaker
+// is activated on that range.
+//
+// To address this problem, all kv operations must be retried if they encounter
```
nit: all KV operations that are a part of the start-up sequence

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,190 @@
+// If your test fails because startup assertion failed, then there's a new code
+// path run during server startup which is not correctly wrapped in retry loop
+// and it could cause node startup failures. You need to identify what is called
+// from Server.PreStart and add retry accordingly.
```
👍🏽 nice comment!

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// RunWithRetry synchronously runs a function with retry of circuit
+// breaker errors on startup.
```
we should also retry AmbiguousResult.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,190 @@
+	// Retry indefinitely until context is cancelled.
+	var result T
+	var err error
+	for r := retry.Start(startupRetryOpts); r.Next(); {
```
Please use `StartWithCtx` so that `r.Next()` doesn't continue blocking once ctx cancels.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,190 @@
+		if ctx.Err() != nil || !IsRetryableReplicaError(err) {
+			break
+		}
+		if r.CurrentAttempt() > retriesBeforeLogging {
```
This is ever so slightly clumsy, how about this:

```go
// at the beginning of this method
every := log.Every(5*time.Second)

for ... {
  // ...
  if every.ShouldLog() {
    log.Infof(...)
  }
}
```

It makes sense to log on the first retry, these retries should be rare and if they exist and become subject of investigation one would like to know the first time they occurred.

This approach also avoids the interdependence of `retriesBeforeLogging` and the retry opts.

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,190 @@
+			log.Infof(ctx, "failed %s during node startup, retrying %s", opName, err)
+		}
+	}
+	// This shouldn't happen as we retry indefinitely.
```
Not true, this is the happy path? And also the path in which ctx gets canceled.

tbg reviewed (changes requested):
Looking pretty good! Two main subjects in this review:

- moving the expensive test to somewhere else and making sure it doesn't run in CI except nightly (or deciding to delete it, let's discuss what incremental value it brings over the fast test you have; should there be a roachtest instead? But that kinda sucks too!)
- retrying ambiguous result errors.

There's also a flurry of smaller comments, but the above are the substantive ones.

aliher1911 committed:
testutil: zone config propagation wait for test cluster

In TestClusters with more than 4 nodes, wait for full replication
doesn't take updated config into account.
This commit adds new check that zone config was propagated prior to
making replication state checks.

Epic: none

Release note: None

aliher1911 committed:
server: retry ambiguous kv/sql operations on node startup

Previously, nodes could fail on startup if node itself was needed to
restore quorum on ranges. This problem was caused by circuit breaker
failing fast on live node when talking back to restarting node instead
of waiting for raft operation to run again (and succeed).
This PR adds explicit retries and also adds assertions enabled in race
build that would fire if startup reaches dist sender or internal sql
executor without wrapping it in startup retry helper.

Release note: None

aliher1911 reviewed (commented):
I'll see how much context improvements could be unimproved.

aliher1911 commented:
For posterity test flakes seem to be caused by https://github.com/cockroachdb/cockroach/issues/99753
Hanging threads match.

tbg commented on pkg/server/server_startup_test.go:
```diff
@@ -0,0 +1,89 @@
+					StickyEngineRegistry: reg,
+				},
+				SpanConfig: &spanconfig.TestingKnobs{
+					ConfigureScratchRange: true,
```
What's this for? Add a comment?

tbg commented on pkg/testutils/soon.go:
```diff
@@ -77,7 +77,7 @@ func SucceedsWithinError(fn func() error, duration time.Duration) error {
 }
 
-func succeedsSoonDuration() time.Duration {
+func SucceedsSoonDuration() time.Duration {
```
Needs comment, isn't the linter yelling at you? If it isn't don't bother.

tbg commented on pkg/ts/db.go:
```diff
@@ -165,15 +166,17 @@ func (db *DB) PollSource(
 			select {
 			case <-ticker.C:
-				p.poll()
+				p.poll(p.AnnotateCtx(context.Background()))
```
why not `bgCtx`?

tbg commented on pkg/upgrade/upgrademanager/manager.go:
```diff
@@ -196,7 +197,7 @@ func (m *Manager) RunPermanentUpgrades(ctx context.Context, upToVersion roachpb.
 		// minimized and tests continue to be happy as they were before the
 		// permanent migrations were introduced.
-		return m.runPermanentMigrationsWithoutJobsForTests(ctx, user)
+		return m.runPermanentMigrationsWithoutJobsForTests(startup.WithoutChecks(ctx), user)
```
Comment on the `WithoutChecks`? Is it because this is already in a retry loop?

tbg commented on pkg/util/startup/retry.go:
```diff
@@ -0,0 +1,205 @@
+}
+
+// IsRetryableReplicaError returns true for replica availability errors.
+func IsRetryableReplicaError(err error) bool {
```
How does this work with SQL queries at start-up? Are we getting the `*kvpb.Error` back in all cases instead of a mangled/flattened SQL error?

tbg commented on pkg/util/startup/startup_test.go:
```diff
@@ -0,0 +1,281 @@
+func TestStartupFailureRandomRange(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.UnderStressRace(t, "6 nodes with replication is too slow for stress race")
```
// This test takes 30s and so we don't want it to run in the "blocking path" of CI at all, and we also don't want to stress it in nightlies as part of a big package (where it will take a lot of time that could be spent running "faster" tests). In this package, it is the only test and so it's fine to run it under nightly (skipping nightly stressrace because race builds with many nodes are very resource intensive and tend to collapse).

tbg reviewed (approved):
Thanks for following through!

aliher1911 commented:
bors r=tbg

aliher1911 commented:
bors r=tbg
